flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java

```diff
@@ -0,0 +1,247 @@
+import org.apache.flink.runtime.highavailability.HighAvailabilityServices;
+import org.apache.flink.runtime.highavailability.HighAvailabilityServicesUtils;
+import org.apache.flink.runtime.metrics.MetricRegistry;
+import org.apache.flink.runtime.metrics.MetricRegistryConfiguration;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.rpc.RpcService;
+import org.apache.flink.runtime.rpc.akka.AkkaRpcService;
+import org.apache.flink.runtime.security.SecurityContext;
+import org.apache.flink.runtime.security.SecurityUtils;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.FlinkException;
+
+import akka.actor.ActorSystem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+
+import scala.concurrent.duration.FiniteDuration;
+
+/**
+ * Base class for the Flink cluster entry points.
+ *
+ * <p>Specialization of this class can be used for the session mode and the per-job mode
+ */
+public abstract class ClusterEntrypoint implements FatalErrorHandler {
+
+	protected static final Logger LOG = LoggerFactory.getLogger(ClusterEntrypoint.class);
+
+	protected static final int SUCCESS_RETURN_CODE = 0;
+	protected static final int STARTUP_FAILURE_RETURN_CODE = 1;
+	protected static final int RUNTIME_FAILURE_RETURN_CODE = 2;
+
+	/** The lock to guard startup / shutdown / manipulation methods. */
+	private final Object lock = new Object();
+
+	@GuardedBy("lock")
+	private MetricRegistry metricRegistry = null;
+
+	@GuardedBy("lock")
+	private HighAvailabilityServices haServices = null;
+
+	@GuardedBy("lock")
+	private BlobServer blobServer = null;
+
+	@GuardedBy("lock")
+	private HeartbeatServices heartbeatServices = null;
+
+	@GuardedBy("lock")
+	private RpcService commonRpcService = null;
+
+	protected void startCluster(String[] args) {
+		final ClusterConfiguration clusterConfiguration = parseArguments(args);
+
+		final Configuration configuration = loadConfiguration(clusterConfiguration);
+
+		try {
+			SecurityContext securityContext = installSecurityContext(configuration);
+
+			securityContext.runSecured(new Callable<Void>() {
+				@Override
+				public Void call() throws Exception {
+					runCluster(configuration);
+
+					return null;
+				}
+			});
+		} catch (Throwable t) {
+			LOG.error("Cluster initialization failed.", t);
+
+			try {
+				shutDown(false);
+			} catch (Throwable st) {
+				LOG.error("Could not properly shut down cluster entrypoint.", st);
+			}
+
+			System.exit(STARTUP_FAILURE_RETURN_CODE);
+		}
+	}
+
+	protected ClusterConfiguration parseArguments(String[] args) {
+		ParameterTool parameterTool = ParameterTool.fromArgs(args);
+
+		final String configDir = parameterTool.get("configDir", "");
+
+		return new ClusterConfiguration(configDir);
+	}
+
+	protected Configuration loadConfiguration(ClusterConfiguration clusterConfiguration) {
```

Comment:
Yes we should add support for dynamic properties. We can add the parsed dynamic properties after we have loaded the configuration. I would like to do it as a follow up because I would also like to change the usage of static fields in `GlobalConfiguration`.